pkg/build/builder/dockerutil.go

```diff
@@ -113,3 +118,60 @@ func buildImage(client DockerClient, dir string, dockerfilePath string, noCache
 	}
 	return client.BuildImage(opts)
 }
+
+// dockerRun mimics the 'docker run' command. It uses the Docker Remote API to
+// create and start a container and stream its logs to stdout and stderr. The
+// container is removed after it terminates.
+func dockerRun(client DockerClient, createOpts docker.CreateContainerOptions, logsOpts docker.LogsOptions) error {
+	// Create a new container.
+	glog.V(4).Infof("Creating container ...")
+	c, err := client.CreateContainer(createOpts)
+	if err != nil {
+		return err
+	}
+
+	// Container was created, so we defer its removal.
+	var rmErr error
+	defer func() {
+		glog.V(4).Infof("Removing container %q ...", c.ID)
+		rmErr = client.RemoveContainer(docker.RemoveContainerOptions{
+			ID: c.ID,
+		})
+		if rmErr == nil {
+			glog.V(4).Infof("Removed container %q", c.ID)
+		} else {
+			glog.V(4).Infof("Failed to remove container %q: %v", c.ID, rmErr)
+		}
+	}()
+
+	// Start the container.
+	glog.V(4).Infof("Starting container %q ...", c.ID)
+	if err := client.StartContainer(c.ID, nil); err != nil {
+		return err
+	}
+
+	// Stream container logs.
+	glog.V(4).Infof("Streaming logs of container %q ...", c.ID)
+	logsOpts.Container = c.ID
+	if err := client.Logs(logsOpts); err != nil {
+		return err
+	}
+
+	// Return an error if the exit code of the container is non-zero.
+	glog.V(4).Infof("Waiting for container %q to stop ...", c.ID)
+	exitCode, err := client.WaitContainer(c.ID)
+	if err != nil {
+		return err
+	}
+	if exitCode != 0 {
+		return fmt.Errorf("container returned non-zero exit code: %d", exitCode)
+	}
+
+	// TODO should we return rmErr or nil? Returning nil means that failing
+	// to remove the container does not report a failure wrt running the
+	// post commit hook (which at this point is guaranteed to have been run
+	// successfully), which might be desirable. The only way to recover from
+	// a failure to remove the container is retrying here or doing some kind
+	// of Docker cleanup in a separate process.
```

Comment:
this runs in separate container, so no. no way this will appear in server log. but if this happen, it means something is broken in Docker, so I'm sure the server log will be flooded with messages about that anyway ;-)